### PR TITLE
Add an `fn_ptr` field to `MiriFrame`

### PR DESCRIPTION
**Pull request by Aaron1011**
The `backtrace-rs` crate can use this to implement
`Frame::symbol_address`, which is used to skip frames
above the call to `Backtrace::capture` on the stack.

The function pointer will not be useable for comparison purposes if the
function is generic, as CTFE creates a new function pointer for each
cast of a (monomorphized) generic function. However, this already
affects code running under Miri, and isn't a problem for `backtrace-rs`
(which only casts a non-generic function).

I've added logic to allow `MiriFrame` to have either 4 or 5 fields - if
a 5th field is present, we write the function pointer to it.

### CODE REVIEW

File: src/shims/backtrace.rs [hunk 1]
```diff
@@ -92,8 +92,18 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
-            throw_ub_format!("bad declaration of miri_resolve_frame - should return a struct with 4 fields");
+        // Reconstruct the original function pointer,
+        // which we pass to user code.
+        let mut fn_ptr = ptr;
+        fn_ptr.offset = Size::from_bytes(0);
+        let fn_ptr = Scalar::Ptr(fn_ptr);
+
+        let num_fields = dest.layout.layout.fields.count();
+
+        if num_fields != 4 && num_fields != 5 {
```
Comment (RalfJung) [inline src/shims/backtrace.rs:103]:
```suggestion
        if !(4..=5).contains(num_fields) {
```
I recently learned this style and now I like to use it everywhere :D

File: src/shims/backtrace.rs [hunk 2]
```diff
@@ -92,8 +92,18 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+            // Always mention 5 fields, since the 4-field struct is only supported
+            // for backwards compatiblity. New code should declare 5 fields
```
Comment (RalfJung) [inline src/shims/backtrace.rs:105]:
```suggestion
            // for backwards compatiblity. New code should declare 5 fields.
```

Comment (RalfJung) [reply src/shims/backtrace.rs:105]: But OTOH, given that we only just introduced this, is it even worth keeping compatibility with the 4-field case?

If we do keep compatibility, I think the docs should mention that, as it is confusing when docs and code diverge. If we consider the 4-field variant deprecated, I am in favor of just dropping support here and now, since it seems unlikely that anyone is already using this feature. This will break your libbacktrace integration but I assume you wanted to update that anyway with this change?

So basically, I think we should either officially support both 4 and 5 fields, or we should deprecate 4 fields and drop them either right now or within the next few weeks.

Comment (Aaron1011) [reply src/shims/backtrace.rs:105]: I'm fine with dropping the 4-field support sometime after this is merged.

**Comment by Aaron1011** [comment]
@RalfJung: Would you like me to add any kind of deprecation warning for the 4-field case, or should we just remove it later without a warning?

**Comment by RalfJung** [comment]
> @RalfJung: Would you like me to add any kind of deprecation warning for the 4-field case, or should we just remove it later without a warning?

It's probably not worth it given the low usage. But please add a comment in the code saying that this is deprecated and slated for removal. You say it is for "backwards compatibility", but that might well mean "we will support this forever due to backwards compatibility".

**Comment by RalfJung** [comment]
Thanks for the PR!
@bors r+
